### PR TITLE
Support static output

### PR DESCRIPTION
**Pull request by 928PJY**
1. change output url schema according to this [Spec](https://github.com/dotnet/docfx/pull/5815)
2. generate toc.html for static putout mode

TODOs: 
1. Handle `redirection-url-not-found` warning for static-output mode
2. Report a warning if the liquid does not exist
3. Whether need to copy resource should be determined by OutputUrlType

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5806)

### CODE REVIEW

File: src/docfx/build/toc/BuildTableOfContents.cs [hunk 1]
```diff
@@ -53,16 +53,23 @@ public static List<Error> Build(Context context, Document file)
             {
                 if (!context.Config.DryRun)
                 {
-                    if (context.Config.Legacy)
+                    if (context.Config.Legacy || context.Config.StaticOutput)
                     {
                         var output = context.TemplateEngine.RunJavaScript("toc.json.js", JsonUtility.ToJObject(model));
-                        context.Output.WriteJson(outputPath, output);
+                        context.Output.WriteJson(LegacyUtility.ChangeExtension(outputPath, ".json"), output);
```
Comment (yufeih) [inline src/docfx/build/toc/BuildTableOfContents.cs:59]: what is this change for?

File: src/docfx/config/Config.cs [hunk 2]
```diff
@@ -60,9 +60,9 @@ internal class Config : PreloadConfig
         public Dictionary<string, GroupConfig> Groups { get; } = new Dictionary<string, GroupConfig>();
 
         /// <summary>
-        /// Gets whether to output JSON model.
+        /// Gets whether to generate static HTML output
         /// </summary>
-        public bool OutputJson { get; private set; } = false;
```
Comment (herohua) [inline src/docfx/config/Config.cs:65]: >OutputJson [](start = 20, length = 10)

this also seems like a breaking change to docs-pipeline - please make sure you have plans for corresponding updates

Comment (928PJY) [reply src/docfx/config/Config.cs:65]: docs-pipeline haven't used those two configs for now, and the default value is still the same with before, now extra work.

Comment (928PJY) [reply src/docfx/config/Config.cs:65]: docs-pipeline haven't used those two configs for now, and the default value is still the same with before, now extra work.

---
In reply to: [411904171](https://github.com/dotnet/docfx/pull/5806#discussion_r411904171) [](ancestors = 411904171)

File: src/docfx/build/toc/BuildTableOfContents.cs (hunk 1 above)
Comment (928PJY) [reply src/docfx/build/toc/BuildTableOfContents.cs:59]: the extension of toc outputPath can be `.html` when the outputType is `Html`

---
In reply to: [411892525](https://github.com/dotnet/docfx/pull/5806#discussion_r411892525) [](ancestors = 411892525)

File: src/docfx/build/document/DocumentProvider.cs [hunk 3]
```diff
@@ -113,94 +123,91 @@ private Document GetDocumentCore(FilePath path)
...
-                        return Path.Combine(Path.GetDirectoryName(path) ?? "", fileName, "index.html").Replace('\\', '/');
-                    }
-                    return Path.ChangeExtension(path, ".json");
-                case ContentType.TableOfContents:
-                    return Path.ChangeExtension(path, ".json");
-                default:
-                    return path;
+                (_, monikers) = _monikerProvider.GetFileLevelMonikers(filePath);
             }
+            catch (Exception ex) when (DocfxException.IsDocfxException(ex, out var dex))
```
[the remark this answers is not in the document]
Comment (928PJY) [reply src/docfx/build/document/DocumentProvider.cs:151]: or the exception will make the main process crashed.

---
In reply to: [413482781](https://github.com/dotnet/docfx/pull/5806#discussion_r413482781) [](ancestors = 413482781)

[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/build/document/DocumentProvider.cs:150]: what is the exception here? should it be `Error` in the first place?

[the remark this answers is not in the document]
Comment (928PJY) [reply src/docfx/build/document/DocumentProvider.cs:150]: move the try catch to the function caller.

---
In reply to: [413496806](https://github.com/dotnet/docfx/pull/5806#discussion_r413496806) [](ancestors = 413496806)

File: docs/specs/markdown.yml [hunk 4]
```diff
@@ -628,15 +628,17 @@ outputs:
 # Liquid template file missing should not crash
```
Comment (yufeih) [inline docs/specs/markdown.yml:628]: consider move these 2 tests to `static.yml`.

File: docs/specs/markdown.yml [hunk 5]
```diff
@@ -628,15 +628,17 @@ outputs:
...
   docs/a.md: Hello `docfx`! 
 outputs: 
   docs/a/index.html:
 ---
 # Source file name ending with spaces should not crash
 inputs:
   docfx.yml: |
-    outputJson: false
+    outputType: html
+    outputUrlType: pretty
```
Comment (yufeih) [inline docs/specs/markdown.yml:641]: what does the output look like for `index.md` when `pretty`?

File: docs/specs/markdown.yml (hunk 4 above)
Comment (928PJY) [reply docs/specs/markdown.yml:628]: move in another PR

---
In reply to: [413546264](https://github.com/dotnet/docfx/pull/5806#discussion_r413546264) [](ancestors = 413546264)

File: docs/specs/markdown.yml (hunk 5 above)
Comment (928PJY) [reply docs/specs/markdown.yml:641]: handled now, see the first test case

---
In reply to: [413550133](https://github.com/dotnet/docfx/pull/5806#discussion_r413550133) [](ancestors = 413550133)

File: src/docfx/build/redirection/RedirectionProvider.cs [hunk 6]
```diff
@@ -218,7 +219,17 @@ private static IEnumerable<string> ProbeRedirectionFiles(string docsetPath, Repo
         {
             var fileUrls = new ConcurrentBag<(FilePath file, string url)>();
             var allSources = _buildScope.Files.Concat(redirectUrlSources);
-            Parallel.ForEach(allSources, file => fileUrls.Add((file, _documentProvider.GetDocument(file).SiteUrl)));
+            Parallel.ForEach(allSources, file =>
```
[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/build/redirection/RedirectionProvider.cs:222]: Not updated?

**Pull request merged**